app/models/user.rb

```diff
@@ -1,24 +1,28 @@
 class User < ActiveRecord::Base
-  has_many :article, dependent: :destroy
-  has_many :comment
-  belongs_to :role
+  has_many :articles, dependent: :destroy, inverse_of: :user
+  has_many :comment, inverse_of: :user
+  belongs_to :role, inverse_of: :users
+
+  ##### ALL VALIDATION
+    ##### Unique email and populated.
 
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, omniauth_providers: [:facebook]
 
   before_create :set_default_role
 
   def is_admin?
-    persisted? and (!role.nil?) and (role.name == "admin") 
+    persisted? and role and (role.name == Role::ADMIN) 
   end
 
   def self.from_omniauth(auth)
-    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
-      user.email = auth.info.email
-      user.password = Devise.friendly_token[0,20]
-    end
+    where('(provider = ? and uid = ?) or lower(email) = ?' , auth.provider, auth.uid, auth.info.email.downcase).last || new
+    #where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
```

Comment:
remove commented code. the old version will always exist in this PR and the commit history